### ISSUE TITLE
Libraries need to be extracted from webview.apk in some cases

### ISSUE THREAD

**Issue by SebiderSushi** [Magisk-Modules-Repo/bromitewebview#13]
On some devices, the libs need to be extracted from the apk file for a succesful install.

Had several devices where webview applications would not work right after installing them via this Magisk Module. After reinstalling the webview.apk file via the system package installer, everything would work. A look into bromite webviews directory under /data/app/ reveals, that the libs folder from within the apk had been extracted within the install dir.

IIRC i have an arm64 device where installing Bromite Webview 78.0.3904.119 had even led to a bootloop, but since i'm currently using the device, i won't try to reproduce it.

So in order to fix this, the libraries from the `libs/` directiory within the webview.apk zip file need to be extracted to `/system/app/webview/libs/arm/` and `/system/app/webview/libs/arm64/` respectively, just like the system package installer does it.
Similar steps are probably required for x86 devices.

**Comment by linuxandria** [comment]
> On some devices, the libs need to be extracted from the apk file for a succesful install.
> 
> Had several devices where webview applications would not work right after installing them via this Magisk Module. After reinstalling the webview.apk file via the system package installer, everything would work. A look into bromite webviews directory under /data/app/ reveals, that the libs folder from within the apk had been extracted within the install dir.
> 
> IIRC i have an arm64 device where installing Bromite Webview 78.0.3904.119 had even led to a bootloop, but since i'm currently using the device, i won't try to reproduce it.
> 
> So in order to fix this, the libraries from the `libs/` directiory within the webview.apk zip file need to be extracted to `/system/app/webview/libs/arm/` and `/system/app/webview/libs/arm64/` respectively, just like the system package installer does it.
> Similar steps are probably required for x86 devices.

Boot scripts resolve this by installing the app manually. Some devices don't reliably run the boot scripts which I've yet to pin down why

**Comment by SebiderSushi** [comment]
Ah, of course. Then, my `service-verbose.log` could be helpful after all:

```
+ set -euo pipefail
+ trap 'exxit $?' EXIT
+ FINDLOG=/data/adb/modules/bromitewebview/logs/find.log 
+ VERBOSELOG=/data/adb/modules/bromitewebview/logs/service-verbose.log 
+ touch /data/adb/modules/bromitewebview/logs/service-verbose.log
+ touch /data/adb/modules/bromitewebview/logs/find.log
+ date
+ echo 'Started at Sun Mar  1 22:05:21 CET 2020'
+ '[' -f /data/adb/modules/bromitewebview/apk/webview.apk ']'
+ sleep 30
+ pm install /data/adb/modules/bromitewebview/apk/webview.apk
Failure [INSTALL_FAILED_ALREADY_EXISTS: Attempt to re-install com.android.webview without first uninstalling.]
+ exxit 1
+ set +euxo pipefail
/data/adb/modules/bromitewebview/service.sh: abort: not found
```

So is it basically about whether the system detects and "installs" (registers?) the system apk file before or after `service.sh` is run?

In my case, calling `pm install` with the `-r` switch in `service.sh` has led to the desired behaviour.

Reference:
```
pm install: install a single legacy package
pm install-create: create an install session
    -l: forward lock application
    -r: replace existing application
    -t: allow test packages
    -i: specify the installer package name
    -s: install application on sdcard
    -f: install application on internal flash
    -d: allow version code downgrade (debuggable packages only)
    -p: partial application install
    -g: grant all runtime permissions
    -S: size in bytes of entire session
```


Now, this is fine in that it will work after installing, but once a user uses the "uninstall updates" function, the system will be back to an incompletely installed webview.

Also, it would be very nice if only one of the webview.apk files would be necessary (`$MODDIR/apk/webview.apk` seems redundant), since 40 MB can still make a difference on some older devices with low storage.

**Comment by linuxandria** [comment]
The APK is downloaded to sdcard so that in the event the boot script doesn't run its easily accessible.
The APK is copied to the module folder so it can be installed before /sdcard may be properly mounted and decrypted.
Also most people don't uninstall updates to the webview :P

**Comment by linuxandria** [comment]
Also the APK is removed once properly installed

**Comment by darylrobert** [comment]
Having the same issue here.

APK not removed.

The error is exactly the same.

Magisk version: v20.1

Device model + software version: xiaomi.eu_multi_HMNote8_V11.0.4.0.PCOCNXM_v11-9.zip

**Comment by linuxandria** [comment]
> Having the same issue here.
> APK not removed, and lib needs to extracted from the APK file.
> The error is exactly the same, in service-verbose.log
> `/data/adb/modules/bromitewebview/service.sh: abort: not found`
> 
> Magisk version: v20.1
> Device model + software version: xiaomi.eu_multi_HMNote8_V11.0.4.0.PCOCNXM_v11-9.zip

Are you using the latest alpha? https://github.com/linuxandria/bromitewebview/releases

**Comment by darylrobert** [comment]
@linuxandria Gee, I already did extracting of lib by hand. No way I would try it again.

Let me see if I can spare time taking a nandroid backup and try installing bromite system webview again. I don't want things to break randomly as my sole phone.

**Comment by darylrobert** [comment]
[bromitewebview-debug.log](https://github.com/Magisk-Modules-Repo/bromitewebview/files/4364751/bromitewebview-debug.log)

**Comment by SebiderSushi** [comment]
I have tried version `5.0.0_beta1` from https://github.com/linuxandria/bromitewebview/releases.
The `service-verbose.log` fails at the `pm install` line with
```
+ pm install -r /data/adb/modules/bromitewebview/apk/webview.apk
+ exxit 2
+ set +euxo pipefail
/data/adb/modules/bromitewebview/service.sh: abort: not found
```
The file `/data/adb/modules/bromitewebview/apk/webview.apk` exists and is the correct and intact webview apk. It succesfully installs if i try to install it manually via my file manager app.

**Comment by SebiderSushi** [comment]
Now, via ADB shell i get the following:
```
[sebi@pc ~]$ adb shell
santoni:/ $ su
santoni:/ # pm install -r /data/adb/modules/bromitewebview/apk/webview.apk
Success
santoni:/ # /data/adb/modules/bromitewebview/service.sh
/system/bin/sh: /data/adb/modules/bromitewebview/service.sh: can't execute: Permission denied
126|santoni:/ # sh /data/adb/modules/bromitewebview/service.sh
Started at Tue Mar 24 04:35:03 CET 2020
cmd: Failure calling service package: Failed transaction (2147483646)
2|santoni:/ # 
```

**Comment by SebiderSushi** [comment]
To pin down the issue i started removing lines from service.sh
Once i uncommented the `exec 2>$MODDIR/logs/service-verbose.log` line, the install went fine.
Then i searched for what `exec` even does - and it seems like `pm` does not like having its streams redirected to a real file.

With the help of [this stackoverflow answer](https://stackoverflow.com/a/25475268), i have found a solution to exclude pm from the stream redirection set by `exec 2>`:
```
exec 2>&-
pm install -r $MODDIR/apk/webview.apk
exec 2>>$LOGFILE
```
With this change the script runs and installs successfully, but it seems as if there must be a cleaner solution to this.


Off-topic: Why are all variables defined three times? Why is `$MODDIR` overwritten? And most improtantly: Why is the indentation of the `exxit()` method so fucked up?

**Comment by SebiderSushi** [comment]
heh, of course there is a better way. now i feel stupid
`pm install -r $MODDIR/apk/webview.apk 2>&-` is enough

**Comment by linuxandria** [comment]
> Now, via ADB shell i get the following:
> 
> ```
> [sebi@pc ~]$ adb shell
> santoni:/ $ su
> santoni:/ # pm install -r /data/adb/modules/bromitewebview/apk/webview.apk
> Success
> santoni:/ # /data/adb/modules/bromitewebview/service.sh
> /system/bin/sh: /data/adb/modules/bromitewebview/service.sh: can't execute: Permission denied
> 126|santoni:/ # sh /data/adb/modules/bromitewebview/service.sh
> Started at Tue Mar 24 04:35:03 CET 2020
> cmd: Failure calling service package: Failed transaction (2147483646)
> 2|santoni:/ # 
> ```
> 
> This device is running LineageOS 16.0

SePolicy issue, I'll have to look into it again

**Comment by linuxandria** [comment]
@SebiderSushi the exxit() method is originally by @JohnFawkes I haven't touched it

**Comment by linuxandria** [comment]
Check 5.0.0_beta2, see if it still persists

**Comment by SebiderSushi** [comment]
When installing `bwv_5.0.0-beta2.zip` in Magisk Manager `7.5.1 (267)` i get the following:
```
- Copying zip to temp directory
- Installing bwv_5.0.0-beta2.zip
Archive:  /data/user/0/com.topjohnwu.magisk/cache/flash/install.zip
  inflating: module.prop
****************************
 Bromite Systemless Webview 
****************************
*******************
 Powered by Magisk 
*******************
Archive:  /data/user/0/com.topjohnwu.magisk/cache/flash/install.zip
  inflating: customize.sh
- Debug mode
  Module install log will include debug info
+ ui_print '- Extracting module files'
+ true
+ echo '- Extracting module files'
+ unzip -o /data/user/0/com.topjohnwu.magisk/cache/flash/install.zip -x 'META-INF/*' common/functions.sh -d /data/adb/modules_update/bromitewebview
  Be sure to save it after module install
- Extracting module files
Archive:  /data/user/0/com.topjohnwu.magisk/cache/flash/install.zip
   creating: common/
   creating: common/addon/
  inflating: common/addon/placeholder
  inflating: common/install.sh
   creating: common/tools/
  inflating: common/tools/curl-arm
  inflating: common/tools/curl-x86
  inflating: common/WebviewOverlay.apk
  inflating: customize.sh
  inflating: LICENSE
  inflating: module.prop
  inflating: post-fs-data.sh
  inflating: sepolicy.rule
  inflating: service.sh
   creating: system/
   creating: system/app/
  inflating: system/app/placeholder
  inflating: uninstall.sh
+ '[' -f /data/adb/modules_update/bromitewebview/common/addon.tar.xz ]
+ ls -A '/data/adb/modules_update/bromitewebview/common/addon/*/install.sh'
+ '['  ]
+ ui_print '- Removing old files'
+ true
+ echo '- Removing old files'
+ '[' -f /data/adb/modules/.bromitewebview-files ]
+ ui_print '- Installing'
+ true
+ echo '- Installing'
+ '[' -f /data/adb/modules_update/bromitewebview/common/install.sh ]
+ . /data/adb/modules_update/bromitewebview/common/install.sh
+ true
+ SDCARD=/storage/emulated/0
+ VERSIONFILE=/sdcard/bromite/version
+ chmod 0755 /data/adb/modules_update/bromitewebview/common/tools/curl-arm
- Removing old files
- Installing
- arm64 SDK 28 system detected, selecting the appropriate files
+ alias 'curl=$MODPATH/common/tools/curl-$ARCH32'
+ ui_print '- arm64 SDK 28 system detected, selecting the appropriate files'
+ true
+ echo '- arm64 SDK 28 system detected, selecting the appropriate files'
+ '[' '!' -f /sdcard/bromite/version ]
+ '[' true '=' true ]
+ mkdir -p /sdcard/bromite
+ ui_print '- Downloading extra files please be patient...'
+ true
+ echo '- Downloading extra files please be patient...'
+ sed -E 's/.*"([^"]+)".*/\1/'
- Downloading extra files please be patient...
+ + grep '"tag_name":'
/data/adb/modules_update/bromitewebview/common/tools/curl-arm -k -L --silent https://api.github.com/repos/bromite/bromite/releases/latest
+ V=81.0.4044.83
+ URL=https://github.com/bromite/bromite/releases/download/81.0.4044.83/arm64_SystemWebView.apk
+ '[' -f /sdcard/bromite/webview.apk ]
+ tr -d .
+ cat /sdcard/bromite/version
+ tr -d .
+ echo 81.0.4044.83
- Extracting webview files...
+ '[' 810404483 -lt 810404483 ]
+ ui_print '- Extracting webview files...'
+ true
+ echo '- Extracting webview files...'
+ '[' '!' -f /sdcard/bromite/webview.apk ]
+ cp_ch /storage/emulated/0/bromite/webview.apk /data/adb/modules_update/bromitewebview/system/app/webview/webview.apk
+ getopt -o nr -- /storage/emulated/0/bromite/webview.apk /data/adb/modules_update/bromitewebview/system/app/webview/webview.apk
+ local 'opt= -- '"'"'/storage/emulated/0/bromite/webview.apk'"'"' '"'"'/data/adb/modules_update/bromitewebview/system/app/webview/webview.apk'"'" 'BAK=true' 'UBAK=true' 'FOL=false'
+ eval set -- ' -- '"'"'/storage/emulated/0/bromite/webview.apk'"'"' '"'"'/data/adb/modules_update/bromitewebview/system/app/webview/webview.apk'"'"
+ set -- -- /storage/emulated/0/bromite/webview.apk /data/adb/modules_update/bromitewebview/system/app/webview/webview.apk
+ true
+ shift
+ break
+ local 'SRC=/storage/emulated/0/bromite/webview.apk' 'DEST=/data/adb/modules_update/bromitewebview/system/app/webview/webview.apk' 'OFILES=/storage/emulated/0/bromite/webview.apk'
+ false
+ '[' -z ]
+ PERM=0644
+ BAK=false
+ false
+ '[' -d /data/adb/modules_update/bromitewebview/system/app/webview/webview.apk ]
+ local 'FILE=/data/adb/modules_update/bromitewebview/system/app/webview/webview.apk'
+ false
+ false
+ install -D -m 0644 /storage/emulated/0/bromite/webview.apk /data/adb/modules_update/bromitewebview/system/app/webview/webview.apk
Segmentation fault
+ ui_print '!!!!!!!!!!!!!!! VERY IMPORTANT PLEASE READ!!!!!!!!!!!!!!!!!'
+ true
+ echo '!!!!!!!!!!!!!!! VERY IMPORTANT PLEASE READ!!!!!!!!!!!!!!!!!'
+ ui_print 'Reboot immediately after flashing or you may experience some issues! '
+ true
+ echo 'Reboot immediately after flashing or you may experience some issues! '
+ ui_print 'Also, if you had any other webview such as Google webview, you may want to re-enable it but beware conflicts'
+ true
+ echo 'Also, if you had any other webview such as Google webview, you may want to re-enable it but beware conflicts'
+ ui_print 'Next boot may take significantly longer, we have to clear Dalvik cache here'
+ true
+ echo 'Next boot may take significantly longer, we have to clear Dalvik cache here'
!!!!!!!!!!!!!!! VERY IMPORTANT PLEASE READ!!!!!!!!!!!!!!!!!
Reboot immediately after flashing or you may experience some issues! 
Also, if you had any other webview such as Google webview, you may want to re-enable it but beware conflicts
Next boot may take significantly longer, we have to clear Dalvik cache here
+ rm -rf '/data/resource-cache/*' '/data/dalvik-cache/*' '/cache/dalvik-cache/*' '/data/*/com.android.webview*' '/data/system/package_cache/*'
+ '[' -d /product/overlay ]
+ '[' -d /vendor/overlay ]
+ mkdir -p /data/adb/modules_update/bromitewebview/system/vendor/overlay
+ cp_ch /data/adb/modules_update/bromitewebview/common/WebviewOverlay.apk /data/adb/modules_update/bromitewebview/system/vendor/overlay
+ getopt -o nr -- /data/adb/modules_update/bromitewebview/common/WebviewOverlay.apk /data/adb/modules_update/bromitewebview/system/vendor/overlay
+ local 'opt= -- '"'"'/data/adb/modules_update/bromitewebview/common/WebviewOverlay.apk'"'"' '"'"'/data/adb/modules_update/bromitewebview/system/vendor/overlay'"'" 'BAK=true' 'UBAK=true' 'FOL=false'
+ eval set -- ' -- '"'"'/data/adb/modules_update/bromitewebview/common/WebviewOverlay.apk'"'"' '"'"'/data/adb/modules_update/bromitewebview/system/vendor/overlay'"'"
+ set -- -- /data/adb/modules_update/bromitewebview/common/WebviewOverlay.apk /data/adb/modules_update/bromitewebview/system/vendor/overlay
+ true
+ shift
+ break
+ local 'SRC=/data/adb/modules_update/bromitewebview/common/WebviewOverlay.apk' 'DEST=/data/adb/modules_update/bromitewebview/system/vendor/overlay' 'OFILES=/data/adb/modules_update/bromitewebview/common/WebviewOverlay.apk'
+ false
+ '[' -z ]
+ PERM=0644
+ BAK=false
+ false
+ '[' -d /data/adb/modules_update/bromitewebview/system/vendor/overlay ]
+ basename /data/adb/modules_update/bromitewebview/common/WebviewOverlay.apk
+ local 'FILE=/data/adb/modules_update/bromitewebview/system/vendor/overlay/WebviewOverlay.apk'
+ false
+ false
+ install -D -m 0644 /data/adb/modules_update/bromitewebview/common/WebviewOverlay.apk /data/adb/modules_update/bromitewebview/system/vendor/overlay/WebviewOverlay.apk
Segmentation fault
update-binary: /data/adb/modules_update/bromitewebview/common/install.sh: line 74: syntax error: unexpected end of file (expecting "fi")
! Installation failed
```
Afterwards, the $MODDIR in modules_update seems to be intact (it also includes the webview.apk) except that it's missing the `$MODDIR/apk/` directory. However, there no $MODDIR is created under `modules` and i can't locate any logfiles both in $MODDIR and in `/sdcard` or `/sdcard/bromite`.

Tested on two different devices with distinct ROM distrubutions.

**Comment by linuxandria** [comment]
Fixed in stable v5.0.1

Boot scripts that install the app may still be broken for some users but you can always install the apk under /sdcard/bromite

**Comment by unixandria-xda** [comment]
Properly fixed in 6.0 since the install over as user app method wasn't working for some
